src/app/obz.service.ts

```diff
@@ -72,123 +61,115 @@ export class ObzService {
-              path: value.toString(),
-              imageData: contents
-            });
-          }));
-        });
+  parseOBZFile(blob: Blob): Promise<OBZBoardSet> {
+    const parseBoard = this.parseBoard;
+    const parseImage = this.parseImage;
+    const parseSound = this.parseSound;
+    const zipper = new JSZip();
+
+    return zipper.loadAsync(blob).then(function(zip) {
+      const manifestFile = zip.file('manifest.json');
+      if (!manifestFile) {
+        throw new FatalOpenVoiceFactoryError(ErrorCodes.MISSING_MANIFEST, 'No manifest file!');
       }
-      Promise.all(promises).then(() => {
-        observer.complete();
+      return manifestFile.async('text').then(function (manifest: string) {
+
+        let manifestJSON = null;
+        try {
+          manifestJSON = JSON.parse(manifest);
+        } catch (error) {
+          throw new FatalOpenVoiceFactoryError(ErrorCodes.MANIFEST_JSON_ERROR, 'manifest.json is not json', error);
+        }
+        const boardSet = new OBZBoardSet();
+        boardSet.rootBoardKey = manifestJSON.root;
+
+        let promises = [];
+        promises = promises.concat(Object.values(manifestJSON.paths.boards).map(board => parseBoard(zip, board.toString(), boardSet)));
+
+        if (manifestJSON.paths.images) {
+          promises = promises.concat(Object.values(manifestJSON.paths.images).map(image => parseImage(zip, image.toString(), boardSet)));
+        }
+
+        if (manifestJSON.paths.sounds) {
+          promises = promises.concat(Object.values(manifestJSON.paths.sounds).map(sound => parseSound(zip, sound.toString(), boardSet)));
+        }
+
+        return Promise.all(promises).then(() => boardSet);
+      }, function (fail) {
+        // error loading manifest
+        throw new FatalOpenVoiceFactoryError(ErrorCodes.MANIFEST_LOAD_ERROR, 'Could not load manifest.json', fail);
       });
-    }
-    return new Observable(loader);
+    }, function (fail) {
+      // error loading zip file
+      throw new FatalOpenVoiceFactoryError(ErrorCodes.ZIP_PARSE_ERROR, 'Could not parse zip file', fail);
+    });
   }
 
-  private parseSounds = (zip, sounds): Observable<ParsedSound> => {
-    this.log('Parsing sounds');
-    const log = this.log;
-    function loader(observer: Observer<ParsedSound>) {
-      const promises = [];
-      if (sounds) {
-        Object.entries(sounds).forEach(([key, value]) => {
-          // log(`Sound key ${key} path ${value}`);
-          promises.push(zip.file(value).async('base64').then(function (contents) {
-            observer.next({
-              path: value.toString(),
-              soundData: contents
-            });
-          }));
-        });
-      }
-      Promise.all(promises).then(() => {
-        observer.complete();
-      });
+  private parseImage = (zip, image: string, boardSet: OBZBoardSet): Promise<void> => {
+    const encoding = image.toLowerCase().endsWith('.svg') ? 'text' : 'base64';
+    const ifl = zip.file(image);
```

Comment:
Please can this, sf and bf all be made into readbale names :)

Reply:
I'm going to assume you mean meaningful and don't just wont iffle, biffle and siffle